packages/devtools_app/lib/src/charts/treemap.dart

```diff
@@ -0,0 +1,517 @@
+  /// L2 and L3 combined include all treemap nodes after the pivot treemap node.
+  /// A combination of elements are put into L2 and L3 so that
+  /// the aspect ratio of the pivot cell (P) is as close to 1 as it can be.
+
+  /// Example layout:
+  /// ```
+  /// ----------------------
+  /// |      |  P   |      |
+  /// |      |      |      |
+  /// |  L1  |------|  L3  |
+  /// |      |  L2  |      |
+  /// |      |      |      |
+  /// ----------------------
+  /// ```
+  List<TreemapCell> buildTreemap({
+    @required List<TreemapNode> children,
+    @required double width,
+    @required double height,
+    @required double x,
+    @required double y,
+  }) {
+    final isHorizontalRectangle = width > height;
+
+    final totalSize = computeByteSizeForNodes(
+      nodes: children,
+      startIndex: 0,
+      endIndex: children.length - 1,
+    );
+
+    if (children.isEmpty) {
+      return [];
+    }
+
+    // Sort list of children, descending in size.
+    children.sort((a, b) => b.byteSize.compareTo(a.byteSize));
+
+    if (children.length <= 2) {
+      final positionedChildren = <TreemapCell>[];
+      double offset = isHorizontalRectangle ? x : y;
+
+      for (final child in children) {
+        final ratio = child.byteSize / totalSize;
+        positionedChildren.add(
+          TreemapCell(
+            key: UniqueKey(),
+            x: isHorizontalRectangle ? offset : x,
+            y: isHorizontalRectangle ? y : offset,
+            width: isHorizontalRectangle ? ratio * width : width,
+            height: isHorizontalRectangle ? height : ratio * height,
+            onTap: () => cellOnTap(rootNode.childrenMap[child.name]),
+            node: child,
+            levelsVisible: widget.levelsVisible - 1,
+          ),
+        );
+
+        offset += isHorizontalRectangle ? ratio * width : ratio * height;
+      }
+
+      return positionedChildren;
+    }
+
+    final pivotIndex = computePivot(children);
+
+    final pivotDataReference = children[pivotIndex];
+    final pSize = pivotDataReference.byteSize;
+
+    final list1 = children.sublist(0, pivotIndex);
+    final list1Size = computeByteSizeForNodes(
+      nodes: list1,
+      startIndex: 0,
+      endIndex: list1.length - 1,
+    );
+
+    List<TreemapNode> list2 = [];
+    int list2Size = 0;
+    List<TreemapNode> list3 = [];
+    int list3Size = 0;
+
+    // The amount of data we have from pivot + 1 (exclusive)
+    // In another words, if we only put one data in l2, how many are left for l3?
+    // [L1, pivotIndex, data, |d|] d = 2
+    final l3MaxLength = children.length - pivotIndex - 1;
+    int bestIndex = 0;
+    double pivotBestWidth = 0;
+    double pivotBestHeight = 0;
+
+    // We need to be able to put at least 3 elements in l3 for this algorithm.
+    if (l3MaxLength >= 3) {
+      double pBestAspectRatio = double.infinity;
+      // Iterate through different combinations of list2 and list3 to find
```

Comment:
in comments wrap variables and members with [ ] (e.g. [list2])

Reply:
Done